### PR TITLE
Add persist feature

### LINKED ISSUE #154: Add session persistence command and RPC
The livesession model includes a field for marking a session as persistent (preventing GC for any livelessons). However, there is currently no option for modifying this state in a safe way.

We should add and test the ability to mark a session as persistent via the `antictl` tool. The command will have to be implemented, as will the server-side RPC in protobuf.

### PR DESCRIPTION
**Pull request by lara29**
CL to add session persist feature into NRE Labs

Closes #154

### CODE REVIEW

**Comment by lara29** [comment]
@Mierdin draft pr for persistence feature

File: db/driver_inmem.go [hunk 1]
```diff
@@ -508,7 +508,7 @@ func (a *ADMInMem) UpdateLiveSessionPersistence(sc ot.SpanContext, lsID string,
 		return err
 	}
 	a.liveSessions[lsID].Persistent = persistent
-	return nil
+	return a.liveSessions[lsID]
```
Comment (Mierdin) [inline db/driver_inmem.go:511]: Please revert this change; we don't need/want to return a whole liveSession from an update function. These are just meant to return a simple error that the caller can check.

File: cmd/antictl/main.go [hunk 2]
```diff
@@ -126,6 +126,31 @@ func main() {
 					},
 				},
 				// TODO (mierdin): Add command to make a session persistent
```
Comment (Mierdin) [inline cmd/antictl/main.go:128]: Don't forget to remove this :)

File: db/driver_inmem.go (hunk 1 above)
Comment (lara29) [reply db/driver_inmem.go:511]: ok, will update the code

File: cmd/antictl/main.go (hunk 2 above)
Comment (lara29) [reply cmd/antictl/main.go:128]: ;) sure, will remove it in final version

**Comment by lara29** [comment]
@Mierdin Please review the changes, thanks!

File: api/exp/livesessions.go [hunk 3]
```diff
@@ -113,6 +159,6 @@ func liveSessionDBToAPI(dbLS *models.LiveSession) *pb.LiveSession {
 // `db` package's equivalent
 func liveSessionAPIToDB(pbLiveSession *pb.LiveSession) *models.LiveSession {
 	liveSessionDB := &models.LiveSession{}
-	copier.Copy(&pbLiveSession, liveSessionDB)
+	copier.Copy(&liveSessionDB, pbLiveSession)
```
Comment (Mierdin) [inline api/exp/livesessions.go:162]: Good grab!

File: api/exp/livesessions.go [hunk 4]
```diff
@@ -98,6 +126,24 @@ func (s *AntidoteAPI) ListLiveSessions(ctx context.Context, _ *empty.Empty) (*pb
 	return &pb.LiveSessions{Items: lsAPIs}, nil
 }
 
+// UpdateLiveSessionPersistence updates the persistence flag in the session database
+func (s *AntidoteAPI) UpdateLiveSessionPersistence(ctx context.Context, persistence *pb.SessionPersistence) (*empty.Empty, error) {
+	span := ot.StartSpan("api_livesession_persist", ext.SpanKindRPCClient)
+	defer span.Finish()
+
+	_, err := s.Db.GetLiveSession(span.Context(), persistence.SessionID)
```
Comment (Mierdin) [inline api/exp/livesessions.go:134]: UpdateLiveSessionPersistence in `db` actually performs a check for an existing livesession after acquiring a mutex lock. This is done so we don't have any weird race conditions in the time between an initial check and the update.

So, let's avoid this first call to `s.Db.GetLiveSession`, and instead just skip ahead to calling `s.Db.UpdateLiveSessionPersistence`.

File: api/exp/livesessions.go [hunk 5]
```diff
@@ -98,6 +126,24 @@ func (s *AntidoteAPI) ListLiveSessions(ctx context.Context, _ *empty.Empty) (*pb
...
+	if err != nil {
+		return &empty.Empty{}, errors.New("livesession not found")
+	}
+
+	err = s.Db.UpdateLiveSessionPersistence(span.Context(), persistence.SessionID, persistence.Persistent)
+	if err != nil {
+		return &empty.Empty{}, errors.New("Unable to update persistence value in record")
```
Comment (Mierdin) [inline api/exp/livesessions.go:141]: Given that this is more of a "superuser" function, let's just return the `err` that we get from the underlying db function, so the client can see what's going on

File: api/exp/livesessions_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+	_, err := api.CreateLiveSession(ctx, &pb.LiveSession{ID: "abcdef", Persistent: false})
+	ok(t, err)
+
+	ls, err := api.GetLiveSession(ctx, &pb.LiveSession{ID: "abcdef"})
+
+	// Assert namespace exists without error
+	ok(t, err)
+	assert(t, (ls != nil), "")
+
```
Comment (Mierdin) [inline api/exp/livesessions_test.go:62]: Would be nice to also add an assertion to check that the `ID` and `Persist` fields are what you expect before the update, so it's obvious that the update function actually changes things.

File: hack/mocks/mock-config-2.yml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (Mierdin) [inline hack/mocks/mock-config-2.yml:1]: Why this file? Seems identical to the previous one.

File: scheduler/namespaces_test.go [hunk 8]
```diff
@@ -63,3 +63,75 @@ func TestNamespaces(t *testing.T) {
 	})
 
 }
+
+func TestSessionPersistence(t *testing.T) {
```
Comment (Mierdin) [inline scheduler/namespaces_test.go:67]: This is excellent, thank you.

File: hack/mocks/mock-config-2.yml (hunk 7 above)
Comment (lara29) [reply hack/mocks/mock-config-2.yml]: I have removed this and updated the code for fetching the curriculumDir.

File: api/exp/livesessions_test.go (hunk 6 above)
Comment (lara29) [reply api/exp/livesessions_test.go:31]: Added extra checks for ID and Persistent

File: api/exp/livesessions.go (hunk 5 above)
Comment (lara29) [reply api/exp/livesessions.go]: done

File: api/exp/livesessions.go (hunk 4 above)
Comment (lara29) [reply api/exp/livesessions.go]: Okay, yeah in that case this check is unnecessary, i'll remove this

File: api/exp/livesessions.go (hunk 3 above)
Comment (lara29) [reply api/exp/livesessions.go:157]: ty :)

File: api/exp/livesessions_test.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	pb "github.com/nre-learning/antidote-core/api/exp/generated"
+	ot "github.com/opentracing/opentracing-go"
+	// Fake clients
+)
+
+// Helper functions courtesy of the venerable Ben Johnson
+// https://medium.com/@benbjohnson/structuring-tests-in-go-46ddee7a25c
+
+// assert fails the test if the condition is false.
```
[the remark this answers is not in the document]
Comment (lara29) [reply api/exp/livesessions_test.go]: done

**Comment by lara29** [comment]
@Mierdin I've made the changes you requested and also tested it locally. Pls review and let me know if it looks good, thanks!

File: cmd/antictl/main.go [hunk 10]
```diff
@@ -125,7 +126,99 @@ func main() {
...
+						sessionid := c.Args()[1]
+
+						ls, err := client.GetLiveSession(context.Background(), &pb.LiveSession{ID: sessionid})
+						if err != nil {
+							fmt.Println(err)
+							os.Exit(1)
+						}
+
+						if ls.Persistent == persistent {
+							fmt.Printf("Persistent state is already %v, returning", persistent)
```
Comment (Mierdin) [inline cmd/antictl/main.go:160]:
```suggestion
							fmt.Printf("Persistent state is already %v, returning\n", persistent)
```

File: cmd/antictl/main.go [hunk 11]
```diff
@@ -125,7 +126,99 @@ func main() {
...
+						}
+
+						_, err = client.UpdateLiveSessionPersistence(context.Background(), &pb.SessionPersistence{SessionID: sessionid, Persistent: persistent})
+
+						if err != nil {
+							fmt.Println(err)
+							os.Exit(1)
+						}
+
+						fmt.Printf("Persistent flag updated for session %s %v", sessionid, persistent)
```
Comment (Mierdin) [inline cmd/antictl/main.go:171]:
```suggestion
						fmt.Printf("Persistent flag updated for session %s %v\n", sessionid, persistent)
```

File: cmd/antictl/main.go [hunk 12]
```diff
@@ -125,7 +126,99 @@ func main() {
...
+						}
+						defer conn.Close()
+						client := pb.NewLiveSessionsServiceClient(conn)
+
+						if len(c.Args()) < 2 {
+							fmt.Println("Missing args to command : antictl livesession persist <true/false> <session id>")
+							os.Exit(1)
+						}
+
+						persistent, err := strconv.ParseBool(c.Args()[0])
```
Comment (Mierdin) [inline cmd/antictl/main.go:146]: Just a UX concern - the current order of the arguments seems a bit clunky:

```
antictl livesession persist true abcdef
```

Given the approach of the current function, I think the best way of doing it would be to reverse the two args so that you provide the session ID first.

```
antictl livesession persist abcdef true
```

File: cmd/antictl/main.go (hunk 10 above)
Comment (lara29) [reply cmd/antictl/main.go]: Agreed, changed the order of args and submitted

File: cmd/antictl/main.go (hunk 12 above)
Comment (lara29) [reply cmd/antictl/main.go]: Agreed, changed the order of args and submitted

File: cmd/antictl/main.go (hunk 11 above)
Comment (lara29) [reply cmd/antictl/main.go]: added the newline and submitted change

File: cmd/antictl/main.go (hunk 10 above)
Comment (lara29) [reply cmd/antictl/main.go]: added the newline and submitted change

**Comment by lara29** [comment]
@Mierdin updated the code as per your comments, pls review, ty

**Comment by Mierdin** [comment]
Excellent, thank you!

**Comment by lara29** [comment]
thanks @Mierdin
